[css-anchor-position-1] height in container query fails to cycle back to trigger next fallback

tuankiet65 opened the pull request:
#### 5912cc91b207c62ed94909aba62726ff89a35caf
<pre>
[css-anchor-position-1] height in container query fails to cycle back to trigger next fallback
<a href="https://rdar.apple.com/154355605">rdar://154355605</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295014">https://bugs.webkit.org/show_bug.cgi?id=295014</a>

Reviewed by Antti Koivisto.

When trying a position-try fallback, we forget to reset the query container
state, so the style resolution machinery don&apos;t re-evaluate container queries
after the new fallback style is applied.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-container-query-expected.txt:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::updateStateForQueryContainer):
(WebCore::Style::TreeResolver::tryChoosePositionOption):

Canonical link: <a href="https://commits.webkit.org/298999@main">https://commits.webkit.org/298999@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/685d93a57440488d7945aaa2dfb0e156e81d4d08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117379 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37049 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123478 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69366 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e3f39d83-5065-4f27-9176-9d3e3c541492) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119257 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37745 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45636 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89070 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43737 "An unexpected error occured. Recent messages:Printed configuration") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30036 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105235 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69577 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b4141ede-5b16-474f-b4b9-be9086c32735) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29094 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23350 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67151 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99439 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126599 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44276 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33271 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97739 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44633 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101469 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97533 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24840 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42892 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20825 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40626 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44149 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49808 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43605 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46950 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45301 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1643,6 +1643,22 @@ std::optional<ResolvedStyle> TreeResolver::tryChoosePositionOption(const Styleab
+        // container queries for pseudo-elements.
+        // FIXME: pseudo-elements can't go with container queries, right?
+        if (styleable.pseudoElementIdentifier)
+            return;
```
Pseudo-elements can be styled via container queries so I suppose the question here is if a pseudo-element can be a container.  The spec says

> For the ::part() and ::slotted() pseudo-element selectors, which represent real elements in the DOM tree, query containers can be established by flat tree ancestors of those elements. For other pseudo-elements, query containers can be established by inclusive flat tree ancestors of their originating element.

"Flat tree ancestors" are always real elements.

tuankiet65 replied:
Yeah what I mean here is whether a pseudo-element can be a query container. Query container/container queries are super easy to get mixed up and they mean entirely different things 😅

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1643,6 +1643,22 @@ std::optional<ResolvedStyle> TreeResolver::tryChoosePositionOption(const Styleab
+        iter->value.invalidated = false;
+    };
+
+    invalidateContainerQuery();
```
Does this really have to happen unconditionally here or could it be done only when an option is actually chosen?

tuankiet65 replied:
I have a patch locally that only invalidates when changing style and it works as well; I'll update it later.

anttijk replied:
I suppose there would be an alternative fix where would make sure not to make any progress in query container state until a position option has been chosen. That is, change the container code instead of position-try code. That would be nicer in a sense that the state wouldn't actually go backwards.

tuankiet65 committed:
[css-anchor-position-1] height in container query fails to cycle back to trigger next fallback
rdar://154355605
https://bugs.webkit.org/show_bug.cgi?id=295014

Reviewed by Antti Koivisto.

When trying a position-try fallback, we forget to reset the query container
state, so the style resolution machinery don't re-evaluate container queries
after the new fallback style is applied.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-container-query-expected.txt:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::updateStateForQueryContainer):
(WebCore::Style::TreeResolver::tryChoosePositionOption):

Canonical link: https://commits.webkit.org/298999@main
